File: Customer/src/main/java/com/postman/alt/service/WorkItemService.java
package com.postman.alt.service;

import com.postman.alt.service.dto.WorkItemActivityResponse;
import com.postman.alt.service.dto.WorkItemCreateRequest;
import com.postman.alt.service.dto.WorkItemResponse;
import com.postman.alt.service.dto.WorkItemUpdateRequest;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.UUID;

public interface WorkItemService {

    WorkItemResponse create(UUID projectId, UUID actorId, WorkItemCreateRequest request);

    Page<WorkItemResponse> list(
            UUID projectId, UUID requesterId, UUID statusId, UUID assigneeId, UUID reporterId, UUID sprintId,
            String priority, String q, Pageable pageable
    );

    WorkItemResponse get(UUID id, UUID requesterId);

    WorkItemResponse update(UUID id, UUID actorId, WorkItemUpdateRequest request);

    WorkItemResponse updateStatus(UUID id, UUID actorId, UUID newStatusId);

    WorkItemResponse updateAssignee(UUID id, UUID actorId, UUID newAssigneeId);

    // gated by WORK_ITEM_EDIT (not WORK_ITEM_ASSIGN) - this is a corrective
    // edit to who's on record as having reported the item, not a work
    // assignment, so it doesn't notify anyone and it's logged like any other
    // field edit (title/description/priority/dueDate).
    WorkItemResponse updateReporter(UUID id, UUID actorId, UUID newReporterId);

    // newSprintId may be null - that moves the item back to the backlog.
    // Gated by WORK_ITEM_EDIT, logged to the activity trail, no notification
    // (same rationale as updateReporter - this is planning, not assignment).
    WorkItemResponse updateSprint(UUID id, UUID actorId, UUID newSprintId);

    void delete(UUID id, UUID actorId);

    Page<WorkItemActivityResponse> getActivity(UUID id, UUID requesterId, Pageable pageable);
}

File: client/src/app/core/services/work-item.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';
import { Observable } from 'rxjs';
import { API_BASE_URL } from '../config/api.config';
import { Page } from '../models/common.model';
import {
  WorkItemActivityResponse,
  WorkItemCreateRequest,
  WorkItemFilter,
  WorkItemResponse,
  WorkItemUpdateRequest,
} from '../models/work-item.model';

const BASE = `${API_BASE_URL}/api`;

@Injectable({ providedIn: 'root' })
export class WorkItemService {
  private readonly http = inject(HttpClient);

  create(projectId: string, request: WorkItemCreateRequest): Observable<WorkItemResponse> {
    return this.http.post<WorkItemResponse>(`${BASE}/projects/${projectId}/work-items`, request);
  }

  list(
    projectId: string,
    filter: WorkItemFilter,
    page: number,
    size: number,
    sort = 'createdAt,desc',
  ): Observable<Page<WorkItemResponse>> {
    let params = new HttpParams().set('page', page).set('size', size).set('sort', sort);
    if (filter.statusId) params = params.set('statusId', filter.statusId);
    if (filter.assigneeId) params = params.set('assigneeId', filter.assigneeId);
    if (filter.reporterId) params = params.set('reporterId', filter.reporterId);
    if (filter.sprintId) params = params.set('sprintId', filter.sprintId);
    if (filter.priority) params = params.set('priority', filter.priority);
    if (filter.q) params = params.set('q', filter.q);
    return this.http.get<Page<WorkItemResponse>>(`${BASE}/projects/${projectId}/work-items`, { params });
  }

  get(id: string): Observable<WorkItemResponse> {
    return this.http.get<WorkItemResponse>(`${BASE}/work-items/${id}`);
  }

  update(id: string, request: WorkItemUpdateRequest): Observable<WorkItemResponse> {
    return this.http.patch<WorkItemResponse>(`${BASE}/work-items/${id}`, request);
  }

  updateStatus(id: string, statusId: string): Observable<WorkItemResponse> {
    return this.http.patch<WorkItemResponse>(`${BASE}/work-items/${id}/status`, { statusId });
  }

  updateAssignee(id: string, assigneeId: string | null): Observable<WorkItemResponse> {
    return this.http.patch<WorkItemResponse>(`${BASE}/work-items/${id}/assignee`, { assigneeId });
  }

  updateReporter(id: string, reporterId: string | null): Observable<WorkItemResponse> {
    return this.http.patch<WorkItemResponse>(`${BASE}/work-items/${id}/reporter`, { reporterId });
  }

  updateSprint(id: string, sprintId: string | null): Observable<WorkItemResponse> {
    return this.http.patch<WorkItemResponse>(`${BASE}/work-items/${id}/sprint`, { sprintId });
  }

  delete(id: string): Observable<void> {
    return this.http.delete<void>(`${BASE}/work-items/${id}`);
  }

  getActivity(id: string, page: number, size: number): Observable<Page<WorkItemActivityResponse>> {
    const params = new HttpParams().set('page', page).set('size', size);
    return this.http.get<Page<WorkItemActivityResponse>>(`${BASE}/work-items/${id}/activity`, { params });
  }
}
